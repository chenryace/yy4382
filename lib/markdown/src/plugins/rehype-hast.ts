import type { Root as HastRoot, Node } from "hast";
import type { Processor } from "unified";
import { removePosition as removePositionUtil } from "unist-util-remove-position";

declare module "unified" {
  interface CompileResultMap {
    HastRoot: HastRoot;
  }
}

export function rehypeHast(
  this: Processor,
  { removePosition }: { removePosition: boolean },
) {
  function compiler(tree: Node): HastRoot {
    if (removePosition) {
      removePositionUtil(tree);
    }
    return tree as HastRoot;
  }
  this.compiler = compiler;
}
