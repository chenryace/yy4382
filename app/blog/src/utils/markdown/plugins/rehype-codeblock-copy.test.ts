import { describe, it, expect } from "vitest";
import { rehypeCodeblockCopy } from "./rehype-codeblock-copy";
import { unified } from "unified";
import remarkParse from "remark-parse";
import remarkRehype from "remark-rehype";
import rehypeStringify from "rehype-stringify";
import rehypeShiki from "@shikijs/rehype";
import { rehypeHast } from "./rehype-hast";
import rehypeRaw from "rehype-raw";
import rehypeRemoveComments from "rehype-remove-comments";

describe("rehypeCodeblockCopy", () => {
  it("should add a copy button to code blocks", async () => {
    const md = `
# ABC

\`\`\`ts
console.log("Hello, world!");
\`\`\`
    `;
    const hast = await unified()
      .use(remarkParse)
      .use(remarkRehype)
      .use(rehypeCodeblockCopy)
      .use(rehypeHast, {
        removePosition: true,
      })
      .process(md);
    expect(JSON.stringify(JSON.parse(String(hast)), null, 2)).toMatchSnapshot();
  });
  it("should not add button to non-code blocks", async () => {
    const md = `
# abc
a \`inline\` code block
`;
    const hast = await unified()
      .use(remarkParse)
      .use(remarkRehype)
      .use(rehypeCodeblockCopy)
      .use(rehypeHast, {
        removePosition: true,
      })
      .process(md);
    expect(JSON.stringify(JSON.parse(String(hast)), null, 2)).toMatchSnapshot();
  });
  it("should work with shiki", async () => {
    const md = `
\`\`\`ts
console.log("Hello, world!");
// comment
\`\`\`
    `;
    const hast = await unified()
      .use(remarkParse)
      .use(remarkRehype)
      .use(rehypeRemoveComments)
      .use(rehypeShiki, {
        theme: "catppuccin-macchiato",
      })
      .use(rehypeCodeblockCopy)
      .use(rehypeHast, {
        removePosition: true,
      })
      .process(md);
    expect(JSON.stringify(JSON.parse(String(hast)), null, 2)).toMatchSnapshot();
  });
});
