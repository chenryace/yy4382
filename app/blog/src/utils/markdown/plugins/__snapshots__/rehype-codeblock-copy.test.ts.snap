// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rehypeCodeblockCopy > should add a copy button to code blocks 1`] = `
"{
  "type": "root",
  "children": [
    {
      "type": "element",
      "tagName": "h1",
      "properties": {},
      "children": [
        {
          "type": "text",
          "value": "ABC"
        }
      ]
    },
    {
      "type": "text",
      "value": "\\n"
    },
    {
      "type": "element",
      "tagName": "pre",
      "properties": {
        "class": [
          "copy-code-pre"
        ]
      },
      "children": [
        {
          "type": "element",
          "tagName": "code",
          "properties": {
            "className": [
              "language-ts"
            ]
          },
          "children": [
            {
              "type": "text",
              "value": "console.log(\\"Hello, world!\\");\\n"
            }
          ]
        },
        {
          "type": "element",
          "tagName": "copy-button",
          "properties": {},
          "children": []
        }
      ]
    }
  ]
}"
`;

exports[`rehypeCodeblockCopy > should not add button to non-code blocks 1`] = `
"{
  "type": "root",
  "children": [
    {
      "type": "element",
      "tagName": "h1",
      "properties": {},
      "children": [
        {
          "type": "text",
          "value": "abc"
        }
      ]
    },
    {
      "type": "text",
      "value": "\\n"
    },
    {
      "type": "element",
      "tagName": "p",
      "properties": {},
      "children": [
        {
          "type": "text",
          "value": "a "
        },
        {
          "type": "element",
          "tagName": "code",
          "properties": {},
          "children": [
            {
              "type": "text",
              "value": "inline"
            }
          ]
        },
        {
          "type": "text",
          "value": " code block"
        }
      ]
    }
  ]
}"
`;

exports[`rehypeCodeblockCopy > should work with shiki 1`] = `
"{
  "type": "root",
  "children": [
    {
      "type": "root",
      "children": [
        {
          "type": "element",
          "tagName": "pre",
          "properties": {
            "class": "shiki catppuccin-macchiato copy-code-pre",
            "style": "background-color:#24273a;color:#cad3f5",
            "tabindex": "0"
          },
          "children": [
            {
              "type": "element",
              "tagName": "code",
              "properties": {},
              "children": [
                {
                  "type": "element",
                  "tagName": "span",
                  "properties": {
                    "class": "line"
                  },
                  "children": [
                    {
                      "type": "element",
                      "tagName": "span",
                      "properties": {
                        "style": "color:#CAD3F5"
                      },
                      "children": [
                        {
                          "type": "text",
                          "value": "console"
                        }
                      ]
                    },
                    {
                      "type": "element",
                      "tagName": "span",
                      "properties": {
                        "style": "color:#8BD5CA"
                      },
                      "children": [
                        {
                          "type": "text",
                          "value": "."
                        }
                      ]
                    },
                    {
                      "type": "element",
                      "tagName": "span",
                      "properties": {
                        "style": "color:#8AADF4;font-style:italic"
                      },
                      "children": [
                        {
                          "type": "text",
                          "value": "log"
                        }
                      ]
                    },
                    {
                      "type": "element",
                      "tagName": "span",
                      "properties": {
                        "style": "color:#CAD3F5"
                      },
                      "children": [
                        {
                          "type": "text",
                          "value": "("
                        }
                      ]
                    },
                    {
                      "type": "element",
                      "tagName": "span",
                      "properties": {
                        "style": "color:#A6DA95"
                      },
                      "children": [
                        {
                          "type": "text",
                          "value": "\\"Hello, world!\\""
                        }
                      ]
                    },
                    {
                      "type": "element",
                      "tagName": "span",
                      "properties": {
                        "style": "color:#CAD3F5"
                      },
                      "children": [
                        {
                          "type": "text",
                          "value": ")"
                        }
                      ]
                    },
                    {
                      "type": "element",
                      "tagName": "span",
                      "properties": {
                        "style": "color:#939AB7"
                      },
                      "children": [
                        {
                          "type": "text",
                          "value": ";"
                        }
                      ]
                    }
                  ]
                },
                {
                  "type": "text",
                  "value": "\\n"
                },
                {
                  "type": "element",
                  "tagName": "span",
                  "properties": {
                    "class": "line"
                  },
                  "children": [
                    {
                      "type": "element",
                      "tagName": "span",
                      "properties": {
                        "style": "color:#939AB7;font-style:italic"
                      },
                      "children": [
                        {
                          "type": "text",
                          "value": "// comment"
                        }
                      ]
                    }
                  ]
                }
              ]
            },
            {
              "type": "element",
              "tagName": "copy-button",
              "properties": {},
              "children": []
            }
          ]
        }
      ]
    }
  ]
}"
`;
